crates/sui-framework/sources/display.move

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Defines a Display struct which defines the way an Object
+/// should be displayed. The intention is to keep data as independent
+/// from its display as possible, protecting the development process
+/// and keeping it separate from the ecosystem agreements.
+///
+/// Each of the fields of the Display object should allow for pattern
+/// substitution and filling-in the pieces using the data from the object T.
+module sui::display {
+    use sui::publisher::{is_package, Publisher};
+    use sui::tx_context::TxContext;
+    use std::string::{String, utf8};
+    use sui::vec_map::{Self, VecMap};
+    use sui::object::{Self, ID, UID};
+    use sui::transfer;
+    use sui::event;
+
+    /// For when T does not belong to package in PublisherCap.
+    const ENotOwner: u64 = 0;
+
+    /// The Display object. Defines the way an object should be
+    /// displayed. Display object can only be created and modified with
+    /// a PublisherCap, making sure that the rules are set by the owner
+    /// of the type.
+    ///
+    /// Each of the display properties should support patterns outside
+    /// of the system, making it simpler to customize Display based
+    /// on the property values of an Object.
+    /// ```
+    /// // Example of a display object
+    /// Display<0x...::capy::Capy> {
+    ///  fields:
+    ///    <name, "Capy {{ genes }}">
+    ///    <link, "https://capy.art/capy/{{ id }}">
+    ///    <image, "https://api.capy.art/capy/{{ id }}/svg">
+    ///    <description, "Lovely Capy, one of many">
+    /// }
+    /// ```
+    ///
+    /// Uses only String type due to external-facing nature of the object,
+    /// the property names have a priority over their types.
+    struct Display<phantom T: key> has key {
+        id: UID,
+        /// Contains fields for display. Currently supported
+        /// fields are: name, link, image and description.
+        fields: VecMap<String, String>,
+    }
+
+    /// Event: emitted when a new Display object has been created for type T.
+    /// Type signature of the event corresponds to the type while id serves for
+    /// the discovery.
+    struct DisplayCreated<phantom T: key> has copy, drop {
+        id: ID
+    }
+
+    /// Set a name for the display.
+    /// Eg: `My lovely capy {{genes}}` (for Capy project).
+    entry public fun set_name<T: key>(pub: &Publisher, d: &mut Display<T>, name: String) {
+        assert!(is_package<T>(pub), ENotOwner);
+        vec_map::insert(&mut d.fields, utf8(b"name"), name)
```

Comment:
Nit: constants for `name` etc would be useful, particularly if we have getters too.

Reply:
Agreed. I guess the main thing stopping me from cleaning it up and designing it in a proper way is a lack of certainty on whether we want to go through with this approach. 😕